### PR TITLE
Change | Updating Azure.Identity version to 1.11.2

### PR DESCRIPTION
**Pull request by JRahnama**
Addresses the issue mentioned at #2460 for [CVE-2024-29992](https://github.com/advisories/GHSA-wvxc-855f-jvrv)

### CODE REVIEW

**Review by arellegue — APPROVED** [review]
lgtm

**Comment by cheenamalhotra** [comment]
Based on NuGet's docs, use * instead. See [Use Floating Versions](https://learn.microsoft.com/en-us/nuget/concepts/dependency-resolution#floating-versions)

The [comments in example](https://learn.microsoft.com/en-us/nuget/concepts/package-versioning?tabs=semver20sort#references-in-project-files-packagereference) show that smallest stable versions are resolved with version ranges.

**Comment by SimonCropp** [comment]
this seems to change significantly more than "Updating Azure.Identity version to 1.11.0". it is also assigning upper version limits for many references

**Comment by SimonCropp** [comment]
also. can we get a patch out for this. people should not need to wait for v6 to get this fix

**Comment by JRahnama** [comment]
> also. can we get a patch out for this. people should not need to wait for v6 to get this fix

Sure, it is in our to do list for next patch release. Date yet TBD.

> this seems to change significantly more than "Updating Azure.Identity version to 1.11.0". it is also assigning upper version limits for many references

This is going to change to wild cards. something like 1.11.* or 1.* to pick the latest stable release of the dependency.

**Comment by SimonCropp** [comment]
> This is going to change to floating versions. something like 1.11.* or 1.* to pick the latest stable release of the dependency.

Doesnt nuget always resolve to the lowest transitive? so, in this case, people would still need an explicit reference to Azure.Identity so as to avoid the CVE?

**Comment by JRahnama** [comment]
> > This is going to change to floating versions. something like 1.11.* or 1.* to pick the latest stable release of the dependency.
> 
> Doesnt nuget always resolve to the lowest transitive? so, in this case, people would still need an explicit reference to Azure.Identity so as to avoid the CVE?

Something similar to [this](https://learn.microsoft.com/en-us/nuget/concepts/package-versioning?tabs=semver20sort#references-in-project-files-packagereference) documentation.

**Comment by JRahnama** [comment]
> > > This is going to change to floating versions. something like 1.11.* or 1.* to pick the latest stable release of the dependency.
> > 
> > 
> > Doesnt nuget always resolve to the lowest transitive? so, in this case, people would still need an explicit reference to Azure.Identity so as to avoid the CVE?
> 
> Something similar to [this](https://learn.microsoft.com/en-us/nuget/concepts/package-versioning?tabs=semver20sort#references-in-project-files-packagereference) documentation.

For example if we had 1.* in our Versions.props, when 1.10.3 was marked as deprecated, the next available version would have been picked which is 1.11.0

**Comment by roji** [comment]
@JRahnama @cheenamalhotra as far as I know, it's discouraged to use floating nuget dependency versions in most production software, because it leads to non-deterministic builds. That is, the moment a dependency releases a version, your next CI run automatically uses that, possibly causing failures, performance regressions or whatever. This can make sense in various prototyping or internal development scenarios, but is less recommended for stable, public software.

The alternative mechanism for keeping versions up to date is [dependabot](https://github.com/dependabot), which is an official github feature; a bot submits PRs updating dependency versions whenever new ones become available, and team members merge these as usual. This ensures a git commit for the dependency version change, allowing tracking exactly what changed and when, e.g. when looking at the project history or diagnosing some problem. I'd recommend looking into enabling dependabot - it's widely used in many .NET projects (including EF and other Microsoft .NET projects) and works very well.

Let me know if the above makes sense and if I've missed something specific to SqlClient.

**Comment by JRahnama** [comment]
> @JRahnama @cheenamalhotra as far as I know, it's discouraged to use floating nuget dependency versions in most production software, because it leads to non-deterministic builds. That is, the moment a dependency releases a version, your next CI run automatically uses that, possibly causing failures, performance regressions or whatever. This can make sense in various prototyping or internal development scenarios, but is less recommended for stable, public software.
> 
> The alternative mechanism for keeping versions up to date is [dependabot](https://github.com/dependabot), which is an official github feature; a bot submits PRs updating dependency versions whenever new ones become available, and team members merge these as usual. This ensures a git commit for the dependency version change, allowing tracking exactly what changed and when, e.g. when looking at the project history or diagnosing some problem. I'd recommend looking into enabling dependabot - it's widely used in many .NET projects (including EF and other Microsoft .NET projects) and works very well.
> 
> Let me know if the above makes sense and if I've missed something specific to SqlClient.

Thanks, @roji. 

It appears that our quest to mitigate the rush for hotfix releases when a dependency becomes deprecated doesn't have a viable solution, based on your insights and we're left with no choice but to release hotfixes whenever a dependency version is deprecated.

**Comment by SimonCropp** [comment]
@JRahnama 

> It appears that our quest to mitigate the rush for hotfix releases when a dependency becomes deprecated doesn't have a viable solution, based on your insights, and we're left with no choice but to release hotfixes whenever a dependency version is deprecated.

IMO the fix is to remove the unnecessarily dependency: https://github.com/dotnet/SqlClient/issues/110

**Comment by roji** [comment]
@SimonCropp assuming you meant #1108, I don't think that's related. Even if the Azure stuff were split into a different package for use only by Azure users (which is what #1108 is about), that "SqlClient Azure" package would still have the exact same issue. In other words, this would stop affecting *all* SqlClient users, but it would still affect *some*.

But at the end of the day, it seems reasonable/standard that if/when a vulnerability is found in a dependency of your package, you release a new version of your package referencing a new version - IMHO that's simply how things are supposed to work. I'd definitely not want users of my package to always automatically get the latest patch version of my package's dependencies, since that would expose them to sudden bugs and behavior changes without them having done any change on their side.

**Comment by SimonCropp** [comment]
@roji 

> Even if the Azure stuff were split into a different package for use only by Azure users (which is what https://github.com/dotnet/SqlClient/issues/1108 is about), that "SqlClient Azure" package would still have the exact same issue. In other words, this would stop affecting all SqlClient users, but it would still affect some.

that is true. but the split would enabled the packages to have a different release cadence. it would be more reasonable for the "SqlClient Azure" package  to do a release in response to a release of Azure.Identity. in fact the "SqlClient Azure" package  could be shipped from the Azure.Identity in lock step.

File: tools/props/Versions.props [hunk 1]
```diff
@@ -27,8 +27,8 @@
   </PropertyGroup>
   <!-- NetFx and NetCore project dependencies -->
   <PropertyGroup>
-    <AzureIdentityVersion>1.10.3</AzureIdentityVersion>
-    <MicrosoftIdentityClientVersion>4.56.0</MicrosoftIdentityClientVersion>
+    <AzureIdentityVersion>1.11.0</AzureIdentityVersion>
+    <MicrosoftIdentityClientVersion>4.60.1</MicrosoftIdentityClientVersion>
```
Comment (David-Engel) [inline tools/props/Versions.props:31]: Time to bump these already... 👀 

https://www.nuget.org/packages/Microsoft.Identity.Client/#versions-body-tab

**Comment by AraHaan** [comment]
Or just wildcard the azure dependencies so then every build of SqlClient would pull in the latest Azure package versions.

**Comment by roji** [comment]
@AraHaan please see discussion above about floating versions.

File: tools/specs/add-ons/Microsoft.Data.SqlClient.AlwaysEncrypted.AzureKeyVaultProvider.nuspec [hunk 2]
```diff
@@ -26,25 +26,25 @@ Microsoft.Data.SqlClient.AlwaysEncrypted.AzureKeyStoreProvider.SqlColumnEncrypti
     <dependencies>
       <group targetFramework="net462">
         <dependency id="Microsoft.Data.SqlClient" version="5.1.5" />
-        <dependency id="Azure.Core" version="[1.35.0,2.0.0)" />
+        <dependency id="Azure.Core" version="[1.38.0,2.0.0)" />
```
Comment (DavoudEshtehari) [inline tools/specs/add-ons/Microsoft.Data.SqlClient.AlwaysEncrypted.AzureKeyVaultProvider.nuspec:29]:
```suggestion
        <dependency id="Azure.Core" version="1.38.0" />
```

File: tools/specs/add-ons/Microsoft.Data.SqlClient.AlwaysEncrypted.AzureKeyVaultProvider.nuspec [hunk 3]
```diff
@@ -26,25 +26,25 @@ Microsoft.Data.SqlClient.AlwaysEncrypted.AzureKeyStoreProvider.SqlColumnEncrypti
...
         <dependency id="Azure.Security.KeyVault.Keys" version="[4.5.0,5.0.0)" />
         <dependency id="Microsoft.Extensions.Caching.Memory" version="6.0.1" />
       </group>
       <group targetFramework="net8.0">
         <dependency id="Microsoft.Data.SqlClient" version="5.1.5" />
-        <dependency id="Azure.Core" version="[1.35.0,2.0.0)" />
+        <dependency id="Azure.Core" version="[1.38.0,2.0.0)" />
```
Comment (DavoudEshtehari) [inline tools/specs/add-ons/Microsoft.Data.SqlClient.AlwaysEncrypted.AzureKeyVaultProvider.nuspec:35]:
```suggestion
        <dependency id="Azure.Core" version="1.38.0" />
```

File: tools/specs/add-ons/Microsoft.Data.SqlClient.AlwaysEncrypted.AzureKeyVaultProvider.nuspec [hunk 4]
```diff
@@ -26,25 +26,25 @@ Microsoft.Data.SqlClient.AlwaysEncrypted.AzureKeyStoreProvider.SqlColumnEncrypti
...
         <dependency id="Azure.Security.KeyVault.Keys" version="[4.5.0,5.0.0)" />
         <dependency id="Microsoft.Extensions.Caching.Memory" version="8.0.0" />
       </group>
       <group targetFramework="net6.0">
         <dependency id="Microsoft.Data.SqlClient" version="5.1.5" />
-        <dependency id="Azure.Core" version="[1.35.0,2.0.0)" />
+        <dependency id="Azure.Core" version="[1.38.0,2.0.0)" />
```
Comment (DavoudEshtehari) [inline tools/specs/add-ons/Microsoft.Data.SqlClient.AlwaysEncrypted.AzureKeyVaultProvider.nuspec:41]:
```suggestion
        <dependency id="Azure.Core" version="1.38.0" />
```

File: tools/specs/add-ons/Microsoft.Data.SqlClient.AlwaysEncrypted.AzureKeyVaultProvider.nuspec [hunk 5]
```diff
@@ -26,25 +26,25 @@ Microsoft.Data.SqlClient.AlwaysEncrypted.AzureKeyStoreProvider.SqlColumnEncrypti
...
         <dependency id="Azure.Security.KeyVault.Keys" version="[4.5.0,5.0.0)" />
         <dependency id="Microsoft.Extensions.Caching.Memory" version="6.0.1" />
       </group>
       <group targetFramework="netstandard2.0">
         <dependency id="Microsoft.Data.SqlClient" version="5.1.5" />
-        <dependency id="Azure.Core" version="[1.35.0,2.0.0)" />
+        <dependency id="Azure.Core" version="[1.38.0,2.0.0)" />
```
Comment (DavoudEshtehari) [inline tools/specs/add-ons/Microsoft.Data.SqlClient.AlwaysEncrypted.AzureKeyVaultProvider.nuspec:47]:
```suggestion
        <dependency id="Azure.Core" version="1.38.0" />
```

File: tools/specs/add-ons/Microsoft.Data.SqlClient.AlwaysEncrypted.AzureKeyVaultProvider.nuspec [hunk 6]
```diff
@@ -26,25 +26,25 @@ Microsoft.Data.SqlClient.AlwaysEncrypted.AzureKeyStoreProvider.SqlColumnEncrypti
     <dependencies>
       <group targetFramework="net462">
         <dependency id="Microsoft.Data.SqlClient" version="5.1.5" />
-        <dependency id="Azure.Core" version="[1.35.0,2.0.0)" />
+        <dependency id="Azure.Core" version="[1.38.0,2.0.0)" />
         <dependency id="Azure.Security.KeyVault.Keys" version="[4.5.0,5.0.0)" />
```
Comment (DavoudEshtehari) [inline tools/specs/add-ons/Microsoft.Data.SqlClient.AlwaysEncrypted.AzureKeyVaultProvider.nuspec:30]: This version should be compatible with Versions.props!

**Comment by SimonCropp** [comment]
Azure.Identity  is now at 1.11.2

**Review by txm229 — APPROVED** [review]

**Review by DavoudEshtehari — APPROVED** [review]

**Comment by BlythMeister** [comment]
when is this likely to hit the shelves?

**Pull request merged**